### PR TITLE
Add build system infrastructure.

### PR DESCRIPTION
**Pull request by hoolioh**
# What does this PR do?

Adds a new build system as described in RFC 0002.

# How to test the change?

The builder crate has been excluded from the default members in the workspace so it's not built when 'build' command is issued with no specific package.

```
export LIBDD_OUTPUT_FOLDER=<path_to_release>
cargo build -p builder--features profiling,data-pipeline,symbolizer,crashtracker --release 
```

### CODE REVIEW

File: docs/RFCs/0002-build-process.md
Comment (r1viollet) [inline docs/RFCs/0002-build-process.md:1]: great doc ! thanks

File: docs/RFCs/0002-build-process.md
Comment (r1viollet) [inline docs/RFCs/0002-build-process.md:4]: minor, link was updated

File: .github/workflows/test.yml
Comment (pawelchcki) [inline .github/workflows/test.yml:140]: <3

File: Cargo.toml
Comment (pawelchcki) [inline Cargo.toml:37]: :exploding_head:

File: .github/workflows/test.yml
Comment (ekump) [inline .github/workflows/test.yml:72]: Just curious why this changed?

File: builder/src/bin/dummy.rs
Comment (ekump) [reply builder/src/bin/dummy.rs:1]: +1, what is the purpose of this?

**Review by ekump — APPROVED** [review]
Not my area of expertise, but LGTM. Nice work!

File: builder/Cargo.toml
Comment (hoolioh) [reply builder/Cargo.toml:10]: Not really. Since telemetry and data-pipeline symbols are reexported through profiling I assumed that prof was kind of the default feature. Should we leave it out?

File: builder/src/bin/dummy.rs
Comment (hoolioh) [reply builder/src/bin/dummy.rs:1]: That's the result of abusing from cargo. Cargo needs one of the following elements in the crate manifest: src/lib.rs, src/main.rs or [[bin]]. Since it wasn't a library nor a binary I chose to add a [[bin]] section just in case it could come in handy execute post-actions after the building. However turns out that for [[bin]] sections is mandatory to have a name and it can't be empty so I named it that way because it does nothing. Anyway maybe it could be named post-action (or something similar) although it currently does nothing. WDYT?

File: .github/workflows/test.yml
Comment (hoolioh) [reply .github/workflows/test.yml:72]: Good catch, probably some ill replace command. I'll revert that.

File: build-profiling-ffi.sh
Comment (hoolioh) [reply build-profiling-ffi.sh:1]: Yes at first I wanted to also make the proper changes in libddprof-build but I think I should bring it back and, when the build system is stable enough, do the changes in the build repo and delete them.

File: builder/build/main.rs
Comment (r1viollet) [inline builder/build/main.rs:152]: NIT: should we add defaults and logs on the output behaviour?

File: builder/build/module.rs
Comment (VianneyRuhlmann) [inline builder/build/module.rs:6]: I think it would be useful to have some documentation on this trait, and on how to add a new crate to the build system. Also would it make sense to define a `new` method so that modules have the same interface, or would that be too rigid ?

File: builder/build/main.rs
Comment (r1viollet) [inline builder/build/main.rs:157]: I had a determine path function in the build-common parts to try and have a common way of defining where we should output files.
What do you think of the logic that is available there ?

Comment (hoolioh) [reply builder/build/main.rs:157]: Good idea. I will add it.

File: builder/build/module.rs
Comment (hoolioh) [reply builder/build/module.rs:6]: Agree on the documentation. Could you elaborate a little bit about the reasoning on adding a new method?

File: builder/build/main.rs
Comment (hoolioh) [inline builder/build/main.rs:199]: @r1viollet WDYT?

Comment (r1viollet) [reply builder/build/main.rs:199]: looks good, thanks!

File: Cargo.toml
Comment (bantonsson) [inline Cargo.toml:37]: So this makes `cargo tree` report a number of dependencies as _separate_ instead of the same, ie. `bytes v1.6.0` instead of `bytes v1.6.0 (*)`. I truly don't understand how `cargo` works, and what this setting means for the final artifacts, but have we checked if this affects the sizes?

Comment (bantonsson) [reply Cargo.toml:37]: While looking at the benchmark time changes, I also noticed that this `default-members` thing has other side effects like not running `clippy` on `builder` from the root directory unless you add `--workspace`. What is the real reason for having this?

File: builder/Cargo.toml
Comment (bantonsson) [reply builder/Cargo.toml:10]: This `default` feature for builder is the thing that makes the benchmark times change. Don't ask me to explain how.

File: Cargo.toml
Comment (hoolioh) [reply Cargo.toml:37]: The reason for not including the builder in the default-members was to save time and resources in scenarios that don't need to build the artifacts, like binding projects or test scenarios.

Comment (bantonsson) [reply Cargo.toml:37]: Removing `--workspace` made no difference to the run-time of the benchmark task. It took 9 minutes 53 seconds, and the last one with `--workspace` took 9 minutes and 50 seconds.

File: builder/build/module.rs
Comment (VianneyRuhlmann) [reply builder/build/module.rs:45]: This would make `main.rs` less module specific since adding a module would always be the same 
```
builder.add_module(Box::new(MyModule::new(...));
```
But that would make it more rigid having to pass useless params for some Module and making it more complicated to add new params. So probably not worth it.

File: builder/build/main.rs
Comment (hoolioh) [reply builder/build/main.rs:201]: I'm not sure if I'm following you there. Here `OUT_DIR` (which is always exported by cargo) act as the default directory in absence of `LIBDD_OUTPUT_FOLDER`. 
What do you mean by adding logs? Using `cargo::warning`?

File: builder/Cargo.toml
Comment (hoolioh) [reply builder/Cargo.toml:10]: @morrisonlevi any more thoughts on this?

Comment (morrisonlevi) [reply builder/Cargo.toml:10]: As a contributing member to the profiling subtree, I am happy to not ship it as default. We don't need special treatment, as far as I am aware.

File: builder/build/main.rs
Comment (r1viollet) [reply builder/build/main.rs:201]: Maybe a small doc is what I was looking for. 
When I initially tried running the build, I was unsure of the different knobs I should be trying out.

File: tools/docker/Dockerfile.build
Comment (r1viollet) [inline tools/docker/Dockerfile.build:164]: This is because we need to have everything built.
NIT, in the future we could introduce an `all` target.

**Comment by r1viollet** [comment]
Happy to merge this when you think we have reached parity with today's build. This is already a leap forward! Some of these can be addressed as follow ups.

File: builder/build/main.rs
Comment (hoolioh) [reply builder/build/main.rs:201]: Oh, you're right I include some documentation in the code but totally forgot to include in the README file.

File: tools/docker/Dockerfile.build
Comment (hoolioh) [reply tools/docker/Dockerfile.build:164]: Godd idea. There will be a second stage for the builder so I think it will be a good moment to add this.

**Comment by hoolioh** [comment]
> Happy to merge this when you think we have reached parity with today's build. This is already a leap forward! Some of these can be addressed as follow ups.

@r1viollet during my testing the binaries generated are on par with what we're building now with the scripts. Also the examples are running fine as well as the tests in all the platforms we're currently using. Originally I wanted to modify the gitlab pipeline to use the builder crate but there were some problems with the Windows build. I talked with Gregory about that and we decided that we will make the changes in another separate PR, that's why the build scripts are still in place. Therefore if you're okay with the latest changes I think we can move forward, merge this one and start working on the next iteration.

**Review by bantonsson — APPROVED** [review]
Even the benchmarks approve 😉

**Comment by r1viollet** [comment]
Ok, let's go! Please give a heads up to everyone when merging and we will look out for any potential issues.

**Pull request merged**
